### ISSUE TITLE
Redirecting and managing C stdout/stderr

### ISSUE THREAD

**Issue by jakirkham** [rapidsai/ucx-py#282]
It's common for C/C++ programs to write log information to stdout and/or stderr. This is certainly true of UCX. However this creates a bit of a conundrum when writing a Python extension like ucx-py. The reason being this is typically handled through logging in Python. Further handles stdout/stderr in Python is handled a bit differently due to encoding/decoding of text in the process. This makes the user experience a bit messy as the C/C++ stdout and stderr get dumped somewhere almost arbitrarily. This could be mixed into Jupyter notebook cell output or Jupyter's own output.

To manage this logging a bit better, it would be nice to capture the C/C++ stdout + stderr and redirect it to Python logs. This can be a bit tricky to do in practice for some of the reasons highlighted above, but is possible. There's a nice exposition of how to solve this problem in [this blogpost]( https://eli.thegreenplace.net/2015/redirecting-all-kinds-of-stdout-in-python/ ). Also there is [a small Python library that solves this problem]( https://github.com/minrk/wurlitzer ). We could either roll our own solution or leverage this existing one. Either way it should improve the Python user's experience.

**Comment by madsbk** [comment]
UCX supports logging: https://github.com/openucx/ucx/wiki/Logging

Haven't test it, but apparently it is possible to redirect the log to a file `UCX_LOG_FILE` or _install custom log message handler._

**Comment by jakirkham** [comment]
Indeed. It looks like one can create a function that matches [this `ucs_log_func_t` signature]( https://github.com/openucx/ucx/blob/v1.7.x/src/ucs/debug/log.h#L78-L92 ) and then [add that logger]( https://github.com/openucx/ucx/blob/9544fb9b6fea9e15620dde279e186a8493b555b7/src/ucs/debug/log.h#L153 ). Not sure, but we may also need to [remove]( https://github.com/openucx/ucx/blob/9544fb9b6fea9e15620dde279e186a8493b555b7/src/ucs/debug/log.h#L154 ) any [existing loggers]( https://github.com/openucx/ucx/blob/9544fb9b6fea9e15620dde279e186a8493b555b7/src/ucs/debug/log.h#L155 ).
